ras_party/controllers/account_controller.py

```diff
@@ -0,0 +1,461 @@
+    2. Register the respondent as a new user with the remote OAuth2 service.
+    3. Lookup the case context from the case service
+    4. Lookup the collection exercise from the collection exercise service
+    5. Lookup the survey from the survey service
+    6. Generate an email verification token, and an email verification url
+    7. Invoke the email verification
+    8. Update the database with the new respondent, establishing an association to the business, and an entry
+       in the enrolment table (along with the survey id / survey name)
+    9. Post a case event to the case service to notify that a new respondent has been created
+    10. If something goes wrong after step 1, attempt to perform a compensating action to remove the OAuth2 user
+       (this is currently mocked out as the OAuth2 server doesn't implement an endpoint to achieve this)
+    """
+
+    # Validation, curation and checks
+    expected = ('emailAddress', 'firstName', 'lastName', 'password', 'telephone', 'enrolmentCode')
+
+    v = Validator(Exists(*expected))
+    if 'id' in party:
+        # Note: there's not strictly a requirement to be able to pass in a UUID, this is currently supported to
+        # aid with testing.
+        log.debug("'id' in respondent post message. Adding validation rule IsUuid")
+        v.add_rule(IsUuid('id'))
+
+    if not v.validate(party):
+        raise RasError(v.errors, 400)
+
+    iac = request_iac(party['enrolmentCode'])
+    if not iac.get('active'):
+        raise RasError("Enrolment code is not active.", status_code=400)
+
+    existing = query_respondent_by_email(party['emailAddress'], session)
+    if existing:
+        raise RasError("User with email address {} already exists.".format(party['emailAddress']), status_code=400)
+
+    case_context = request_case(party['enrolmentCode'])
+    case_id = case_context['id']
+    business_id = case_context['partyId']
+    collection_exercise_id = case_context['caseGroup']['collectionExerciseId']
+    collection_exercise = request_collection_exercise(collection_exercise_id)
+
+    try:
+        survey_id = collection_exercise['surveyId']
+        survey = request_survey(survey_id)
+        survey_name = survey['longName']
+    except KeyError:
+        raise RasError("There is no survey bound for this user with email address: {}".format(party['emailAddress']))
+
+    b = query_business_by_party_uuid(business_id, session)
+    if not b:
+        msg = "Could not locate business with id '{}' when creating business association.".format(business_id)
+        raise RasError(msg, status_code=404)
+
+    # Chain of enrolment processes
+    translated_party = {
+        'party_uuid': party.get('id') or str(uuid.uuid4()),
+        'email_address': party['emailAddress'],
+        'first_name': party['firstName'],
+        'last_name': party['lastName'],
+        'telephone': party['telephone'],
+        'status': RespondentStatus.CREATED
+    }
+
+    try:
```

Comment:
Should we separate the process of adding the respondent to the database into a separate function? We have all the queries to the database separate already

Reply:
Not sure it's worth splitting out as of now, nothing else currently uses this functionality.